### PR TITLE
Signing optimizations

### PR DESCRIPTION
**Pull request by thibault-martinez**
- Sponge interface has been removed
- Sponge factory has been removed
- Keccak384 now works with pointers instead of vectors
- Digest method of Keccak384 has been removed (was only used for test purposes)
- Kerl now works with bytes instead of trits
- finalSqueeze method has been added to Kerl
- Signing operations have been drastically optimized avoiding a lot of trits-bytes conversions
- Some bigint methods have been publicly exposed
- Trytes to bytes and vice versa conversions have been added
- Tests have been updated

### CODE REVIEW

File: include/iota/types/big_int.hpp [hunk 1]
```diff
@@ -69,18 +69,20 @@ class Bigint {
    *
    * @return The bytes.
    */
-  std::vector<int8_t> toBytes() const;
+  void toBytes(std::vector<uint8_t> &bytes, std::size_t offset = 0) const;
+
+public:  // TODO(Bigint class or namespace ?)
```
Comment (Cylix) [inline include/iota/types/big_int.hpp]: this comment can be removed

File: include/iota/crypto/kerl.hpp [hunk 2]
```diff
@@ -46,25 +45,33 @@ class Kerl : public ISponge {
...
-   * @param length Length of the current state of the sponge that should be used for squeezing.
+   * @param bytes Bytes to be squeezed based on current state of the sponge.
+   * @param offset Offset at which the bytes should be modified.
    */
-  void squeeze(Types::Trits& trits, std::size_t offset = 0, std::size_t length = 0) override;
+  void squeeze(std::vector<uint8_t>& bytes, std::size_t offset = 0);
+
+  /**
+   * Squeeze the final current state of the sponge to the given trits.
+   * Kerl should not be used after this function is called, unless it is reseted.
```
Comment (Cylix) [inline include/iota/crypto/kerl.hpp]: *reset :D

File: source/crypto/kerl.cpp [hunk 3]
```diff
@@ -40,41 +42,40 @@ Kerl::reset() {
...
-  if (length % TritHashLength != 0)
-    throw Errors::Crypto("Kerl::squeeze failed : illegal length");
-  while (offset < length) {
-    auto bytes = keccak_.squeeze();
-    trits      = Types::bytesToTrits(bytes);
+Kerl::squeeze(std::vector<uint8_t>& bytes, std::size_t offset) {
+  uint8_t state[ByteHashLength];
+  keccak_.squeeze(bytes.data() + offset);
+  std::memcpy(state, bytes.data() + offset, ByteHashLength);
+  Types::Bigint b;  // TODO(bigint class or namespace ?)
```
Comment (Cylix) [inline source/crypto/kerl.cpp]: remove comment here too?

File: source/crypto/kerl.cpp [hunk 4]
```diff
@@ -40,41 +42,40 @@ Kerl::reset() {
...
-    std::transform(bytes.begin(), bytes.end(), bytes.begin(),
-                   [](const int8_t& byte) { return byte ^ 0xFF; });
-    keccak_.reset();
-    keccak_.absorb(bytes);
-    offset += TritHashLength;
-  }
+void
+Kerl::finalSqueeze(std::vector<uint8_t>& bytes, std::size_t offset) {
+  keccak_.squeeze(bytes.data() + offset);
+  Types::Bigint b;  // TODO(bigint class or namespace ?)
```
Comment (Cylix) [inline source/crypto/kerl.cpp]: same here

File: source/crypto/signing.cpp [hunk 5]
```diff
@@ -39,116 +40,110 @@ namespace Signing {
...
-  return addressTrits;
-}
-
+// TODO(Optimize)
 Types::Trits
 signatureFragment(const std::vector<int8_t>& normalizedBundleFragment,
                   const Types::Trits&        keyFragment) {
   Kerl         k;
-  Types::Trits signatureFragment;
+  Types::Trits signatureFragment{};
```
Comment (Cylix) [inline source/crypto/signing.cpp]: not sure initializer is useful here, but you can leave it if you want

**Comment by Cylix** [comment]
Just some syntax/code cleaning changes, otherwise great job, looking forward to merge that and release it!

File: source/crypto/signing.cpp (hunk 5 above)
Comment (thibault-martinez) [reply source/crypto/signing.cpp]: Weird test purposes, will be removed thanks.

**Comment by thibault-martinez** [comment]
Comments have been addressed, thanks.
You can merge when CI reports are up.

**Pull request merged**